openwisp_monitoring/db/backends/influxdb/tests.py

```diff
@@ -240,3 +246,32 @@ def test_metric_write_microseconds_precision(self):
         m.write('00:14:5c:00:00:00', time=datetime(2020, 7, 31, 22, 5, 47, 235142))
         m.write('00:23:4a:00:00:00', time=datetime(2020, 7, 31, 22, 5, 47, 235152))
         self.assertEqual(len(m.read()), 2)
+
+    @patch.object(
+        InfluxDBClient, 'write', side_effect=InfluxDBServerError('Server error')
+    )
+    def test_write_retry(self, mock_write):
+        with self.assertRaises(TimeseriesWriteException):
+            timeseries_db.write('test_write', {'value': 1})
+        m = self._create_general_metric(name='Test metric')
+        with self.assertRaises(Retry):
+            m.write(1)
+
+    @patch.object(timeseries_db, 'write')
+    def test_timeseries_write_params(self, mock_write):
+        m = self._create_general_metric(name='Test metric')
+        with freeze_time('Jan 14th, 2020', auto_tick_seconds=15):
+            # First tick, time should jump by 15 seconds
+            first_time = now()
+            # Two more ticks, metric.write and check_threshold, time should increment by 30s
+            m.write(1)
```

Comment:
We want to be 100% sure, now and in the future, that when a write fails, it's retried (we already test this) **but when the write is retried, the time used for the write is the original one**.

This is the full scenario, but I don't see the retry mechanism involved in this test. So it looks like something important is missing to me.

I think it should be something like the following (**pseudocode alert!**):

```python
@freeze_time('Jan 14th, 2020'):
def test_timeseries_write_params(self, mock_write):
   m = self._create_general_metric(name='Test metric')
   with patch(MAKE_IT_FAIL) as whatever:
       m.write(1)
       # expect retry but find a way to not call the retry

   time_tick_10_minutes()  # do it manually, it's more explicit
   self.assertEqual(now(), first_time + timedelta(minutes=10))

   with patch.object(timeseries_db, 'write') as mock_write:
       trigger_retry()
       mock_write.assert_called_with(
           'test_metric',
           {'value': 1},
           database=None,
           retention_policy=None,
           tags={},
           
           # =================================================================
           # this must be the original time at the moment of the first failure
           # =================================================================
           timestamp='2020-01-14T00:00:00Z',
       )
```